crates/ironrdp-dvc/src/pdu.rs

```diff
@@ -0,0 +1,770 @@
+    fn decode(src: &mut ReadCursor<'_>) -> PduResult<Self> {
+        let header = Header::decode(src)?;
+        match header.cmd {
+            Cmd::Create => Ok(Self::Create(CreateRequestPdu::decode(header, src)?)),
+            Cmd::DataFirst => Ok(Self::Data(DrdynvcDataPdu::DataFirst(DataFirstPdu::decode(
+                header, src,
+            )?))),
+            Cmd::Data => Ok(Self::Data(DrdynvcDataPdu::Data(DataPdu::decode(header, src)?))),
+            Cmd::Close => Ok(Self::Close(ClosePdu::decode(header, src)?)),
+            Cmd::Capability => Ok(Self::Capabilities(CapabilitiesRequestPdu::decode(header, src)?)),
+            _ => Err(unsupported_pdu_err!("Cmd", header.cmd.into())),
+        }
+    }
+}
+
+// Dynamic virtual channel PDU's are sent over a static virtual channel, so they are `SvcPduEncode`.
+impl SvcPduEncode for DrdynvcDataPdu {}
+impl SvcPduEncode for DrdynvcClientPdu {}
+impl SvcPduEncode for DrdynvcServerPdu {}
+
+/// [2.2] Message Syntax
+///
+/// [2.2]: https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpedyc/0b07a750-bf51-4042-bcf2-a991b6729d6e
+#[derive(Debug, PartialEq)]
+pub struct Header {
+    cb_id: FieldType, // 2 bit
+    sp: FieldType,    // 2 bit; meaning depends on the cmd field
+    cmd: Cmd,         // 4 bit
+}
+
+impl Header {
+    pub const FIXED_PART_SIZE: usize = 1;
+    /// Create a new `Header` with the given `cb_id_val`, `sp_val`, and `cmd`.
+    ///
+    /// If `cb_id_val` or `sp_val` is not relevant for a given `cmd`, it should be set to 0 respectively.
+    fn new(cb_id_val: u32, sp_val: u32, cmd: Cmd) -> Self {
+        Self {
+            cb_id: FieldType::for_val(cb_id_val),
+            sp: FieldType::for_val(sp_val),
+            cmd,
+        }
+    }
+
+    fn encode(&self, dst: &mut WriteCursor<'_>) -> PduResult<()> {
+        ensure_size!(in: dst, size: Self::size());
+        dst.write_u8((self.cmd as u8) << 4 | Into::<u8>::into(self.sp) << 2 | Into::<u8>::into(self.cb_id));
+        Ok(())
+    }
+
+    fn decode(src: &mut ReadCursor<'_>) -> PduResult<Self> {
+        ensure_size!(in: src, size: Self::size());
+        let byte = src.read_u8();
+        let cmd = Cmd::try_from(byte >> 4)?;
+        let sp = FieldType::from((byte >> 2) & 0b11);
+        let cb_id = FieldType::from(byte & 0b11);
+        Ok(Self { cb_id, sp, cmd })
+    }
+
+    fn size() -> usize {
+        Self::FIXED_PART_SIZE
+    }
+}
+
+/// [2.2] Message Syntax
+///
+/// [2.2]: https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpedyc/0b07a750-bf51-4042-bcf2-a991b6729d6e
+#[repr(u8)]
+#[derive(Debug, Copy, Clone, PartialEq)]
+enum Cmd {
+    Create = 0x01,
+    DataFirst = 0x02,
+    Data = 0x03,
+    Close = 0x04,
+    Capability = 0x05,
+    DataFirstCompressed = 0x06,
+    DataCompressed = 0x07,
+    SoftSyncRequest = 0x08,
+    SoftSyncResponse = 0x09,
+}
+
+impl TryFrom<u8> for Cmd {
+    type Error = PduError;
+
+    fn try_from(byte: u8) -> Result<Self, Self::Error> {
```

Comment:
For some reason I had it in my mind that we were trying to move away from using those crates and just build things custom, as needed.

cc @CBenoit

Reply:
@pacmancoder Yes, I don’t want to depend on `num_derive` long term

Reply:
@CBenoit I decided an enum was appropriate here because the `Cmd` value is needed to determine how to decode the rest of the message. I think that the API for drdynvc headers is stable. If we get back any of `DataFirstCompressed`, `DataCompressed`, `SoftSyncRequest`, `SoftSyncResponse` then we likely want to end the session and alert the user, who can remediate the issue by removing the dynamic channel client/server implementation that induced one of those messages to be sent. If we receive something outside of that enum, we're really in trouble and it's probably indicative of a fatal bug in the code.